### PR TITLE
Nearly fully documented the optional PE header.

### PR DESCRIPTION
**Pull request by JohnScience**
There's still some work needed for documentation of data directories field in the optional header but it's a separate module.

This PR also...
* introduces two new modules for constants that are used in the optional header (`dll_characteristic` and `subsystem`).
* further refines the documentation of the COFF header module.

### CODE REVIEW

**Comment by JohnScience** [comment]
@m4b

File: src/pe/header.rs [hunk 1]
```diff
@@ -396,73 +407,300 @@ impl Default for DosStub {
...
+    /// [Source](https://0xrick.github.io/win-internals/pe4/#file-header-image_file_header).
+    // Q (JohnScience): Why is the name `number_of_symbol_table` and not `number_of_symbols`?
+    #[doc(alias("NumberOfSymbols"))]
     pub number_of_symbol_table: u32,
+    /// The size of the optional header, in bytes. This value should be zero for object files.
+    ///
+    /// The [`goblin::pe::optional_header::OptionalHeader`](crate::pe::optional_header::OptionalHeader) is meant to
+    /// represent either the 32-bit or the 64-bit optional header. The size of the optional header is used to determine
+    /// which one it is.
+    #[doc(alias("SizeOfOptionalHeader"))]
```
Comment (m4b) [inline src/pe/header.rs:497]: what does this alias?

File: src/pe/header.rs [hunk 2]
```diff
@@ -396,73 +407,300 @@ impl Default for DosStub {
...
     pub characteristics: u16,
 }
 
 pub const SIZEOF_COFF_HEADER: usize = 20;
 /// PE\0\0, little endian
 pub const PE_MAGIC: u32 = 0x0000_4550;
 pub const SIZEOF_PE_MAGIC: usize = 4;
-/// The contents of this field are assumed to be applicable to any machine type
+
+// Q (JohnScience): doesn't it make sense to move all these constants to a dedicated module
```
Comment (m4b) [inline src/pe/header.rs:511]: possibly, yes; it's what we do for other formats iirc

File: src/pe/header.rs [hunk 3]
```diff
@@ -396,73 +407,300 @@ impl Default for DosStub {
...
+// and then re-export them from here? This way, the module will be more organized.
+//
+// Also, don't we want to declare them in a macro to remove the boilerplate and make the implementation
```
Comment (m4b) [inline src/pe/header.rs:514]: file a issue if you feel theres room for improvement :)

File: src/pe/header.rs [hunk 4]
```diff
@@ -531,11 +769,23 @@ impl CoffHeader {
...
+/// The PE header is located at the very beginning of the file and
+/// is followed by the section table and sections.
 #[derive(Debug, PartialEq, Copy, Clone, Default)]
 pub struct Header {
     pub dos_header: DosHeader,
     /// DOS program for legacy loaders
     pub dos_stub: DosStub,
+
+    // Q (JohnScience): should we care about the "rich header"?
+    // https://0xrick.github.io/win-internals/pe3/#rich-header
```
Comment (m4b) [inline src/pe/header.rs:785]: this might be a good thing if it is real and people use it for stuff :shrug:

File: src/pe/optional_header.rs [hunk 5]
```diff
@@ -6,52 +8,125 @@ use crate::pe::data_directories;
...
 pub struct StandardFields {
+    /// The state of the image file. This member can be one of the following values:
+    ///
+    /// * [`IMAGE_NT_OPTIONAL_HDR32_MAGIC`].
+    /// * [`IMAGE_NT_OPTIONAL_HDR64_MAGIC`].
+    /// * [`IMAGE_ROM_OPTIONAL_HDR_MAGIC`].
+    #[doc(alias = "Magic")]
     pub magic: u16,
+    /// The major version number of the linker.
+    #[doc(alias = "MajorLinkerVersion")]
```
Comment (m4b) [inline src/pe/optional_header.rs:102]: i don't think this aliases anything unless i'm misunderstanding what `doc(alias` does

**Review by m4b — COMMENTED** [review]
to be honest this is quite a lot of documentation to review, though not a bad thing, thank you for the PR! What a strange world we're entering with gen ai though :)

my only major concerns are about the various doc aliases, they appear to alias things that don't exist, and i think that should be fixed before merging

File: src/pe/header.rs (hunk 1 above)
Comment (JohnScience) [reply src/pe/header.rs:497]: @m4b When searching in docs for `SizeOfOptionalHeader`, it'll be found.

Here's the quote from the [documentation of `doc(alias = ".."`](https://doc.rust-lang.org/rustdoc/advanced-features.html#add-aliases-for-an-item-in-documentation-search)

> This feature allows you to add alias(es) to an item when using the rustdoc search through the doc(alias) attribute.

Comment (m4b) [reply src/pe/header.rs:497]: ah interesting, i thought it aliased to the doc's of the type/expr in quotes (which doesn't exist); testing it locally, it does indeed work that way, neat, thank you, learned something new today!

File: src/pe/header.rs (hunk 4 above)
Comment (m4b) [reply src/pe/header.rs:785]: @JohnScience would you mind filing an issue for the rich header?

File: src/pe/optional_header.rs (hunk 5 above)
Comment (m4b) [reply src/pe/optional_header.rs:102]: no i was wrong :)

**Review by m4b — APPROVED** [review]
thanks!

**Comment by m4b** [comment]
release: non-breaking

File: src/pe/optional_header.rs (hunk 5 above)
Comment (JohnScience) [reply src/pe/optional_header.rs:102]: Most likely, you were misunderstanding. The link to the explanation provided above.

File: src/pe/header.rs (hunk 3 above)
Comment (JohnScience) [reply src/pe/header.rs:514]: @m4b I will!

File: src/pe/header.rs (hunk 4 above)
Comment (JohnScience) [reply src/pe/header.rs:785]: @m4b It is mostly useful for cybersecurity.

Comment (JohnScience) [reply src/pe/header.rs:785]: > @JohnScience would you mind filing an issue for the rich header?

Done, https://github.com/m4b/goblin/issues/400

**Comment by JohnScience** [comment]
> to be honest this is quite a lot of documentation to review, though not a bad thing, thank you for the PR! What a strange world we're entering with gen ai though :)
> 
> my only major concerns are about the various doc aliases, they appear to alias things that don't exist, and i think that should be fixed before merging

Doc aliases improve the searchability of various items by allowing users finding them by their Windows name. See [Rustdoc > Advanced features > Add aliases for an item in documentation search](https://doc.rust-lang.org/rustdoc/advanced-features.html#add-aliases-for-an-item-in-documentation-search)

**Pull request merged**